lib/liquid/parser_switching.rb

```diff
@@ -1,16 +1,14 @@
 module Liquid
   module ParserSwitching
     def parse_with_selected_parser(markup)
-      case @options[:error_mode] || Template.error_mode
+      case @options.error_mode
       when :strict then strict_parse_with_error_context(markup)
       when :lax    then lax_parse(markup)
       when :warn
         begin
           return strict_parse_with_error_context(markup)
         rescue SyntaxError => e
-          e.set_line_number_from_token(markup)
-          @warnings ||= []
-          @warnings << e
+          @options.warnings << e
```

Comment:
I kept the `@options` instance variable because liquid-c refers to it directly in a [Liquid::Variable#lax_parse monkey patch](https://github.com/Shopify/liquid-c/blob/35e9aee48d639ae1d3ac9ba77616aca9800eab7d/lib/liquid/c.rb#L46).  However, I tried to get rid of direct usage of `@options` in liquid for now, and made sure Tag and Variable both have methods to use instead of `@options` so that liquid-c can be decoupled from a specific instance name for options.